fastjsonschema/generator.py

```diff
@@ -135,8 +156,10 @@ def generate_validation_function(self, uri, name):
         self._validation_functions_done.add(uri)
         self.l('')
         with self._resolver.resolving(uri) as definition:
-            with self.l('def {}(data):', name):
-                self.generate_func_code_block(definition, 'data', 'data', clear_variables=True)
+            self.code_break()
+            with self.l('def {{}}(data, scope="{}"):'.format(self._root_name), name):
```

Comment:
when you have a lot of json schema files e.g. article.json product.json etc, our api had to come up with better error messages that `data.teaser[0] must be ....`, so now when compiling article.json with root_name='article' the message is `article.teaser[0] must be ...`

Reply:
So your API returns in the response the message from this library? I think this can be improved with my idea of more information on the exception class, so your API can use it and generate nice messages which matches the other messages in your app.

Reply:
if you want (as i read in feature requests) to add attributes to the exception - that's great. for me the most important feature is to keep track of the context/scope of where the validation process is at when the exception is thrown. so given: a schema:
article.json
```{
  "$schema": "http://json-schema.org/draft-07/schema#",
  "type" : "object",
  "additionalProperties": false,
  "properties": {
    "teaser" : { "$ref": "intro.json" },
    "prolog" : { "$ref": "intro.json" }
  }
}```
when the error occurs during the prolog validation, the exception contains an attribute with value `article.prolog.....`  about where it happened (like the jsonschema library does).

Reply:
I see. I will make issue for that. Or do you want to do it on top of latest master?

Reply:
well, i suppose now i will have to wait for you to finish your ideas and then try to somehow keep the best of two. as i said, basically i was pretty much content with the refactor and now it is part of the production code. my pull request was merely a try to give back some of my effort for better user experience of your library

Reply:
I have my changes released. I created issue for your needs. If you want, you can upgrade your version of fastjsonschema, apply that feature on top of it and do a new merge request, or I can do it. :-)

Reply:
Well, I would like to contribute if you are interested but as you pointed out your vision of where to go differs from what I have proposed, so perhaps working collaboratively might be a better way than keep merging two ideas all the way. Let me know what you think.

Reply:
You can check out the exceptions--they have more information which can help apps to generate better message for the user. Another step is to be able to return all errors at once. Your improvement is also good, to remember the whole path in connected definitions. If you want to implement this on top of the master, go ahead. If you want to keep your version of JSON schema, it's ok, I can implement it here one day.